File: xanathar/Day02.cs
using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Linq;
using System.Collections.Generic;
using System.IO;
using Helpers;

namespace Day02
{
    enum CommandType
    {
        Forward,
        Depth,
    }

    class Command
    {
        public int Num;
        public CommandType Cmd;

        public static Command Parse(string s)
        {
            string[] parts = s.Split(' ');
            CommandType type = CommandType.Depth;
            int num = int.Parse(parts[1]);

            if (parts[0] == "forward") type = CommandType.Forward;
            else if (parts[0] == "down") type = CommandType.Depth;
            else if (parts[0] == "up") { type = CommandType.Depth; num = -num; }

            return new Command
            {
                Num = num,
                Cmd = type,
            };
        }
    }


    public class Test1 : DayBase
    {
        public override void Run()
        {
            var data = File.ReadAllLines("data/day2.txt").Where(s => !string.IsNullOrEmpty(s)).Select(s => Command.Parse(s)).ToArray();

            int depth = 0, pos = 0;

            foreach (Command c in data)
            {
                if (c.Cmd == CommandType.Forward) pos += c.Num;
                else depth += c.Num;
            }

            Console.WriteLine("{0}p x {1}d = {2}", pos, depth, pos * depth);
        }
    }

    public class Test2 : DayBase
    {
        public override void Run()
        {
            var data = File.ReadAllLines("data/day2.txt").Where(s => !string.IsNullOrEmpty(s)).Select(s => Command.Parse(s)).ToArray();

            int aim = 0, pos = 0, depth = 0;

            foreach (Command c in data)
            {
                if (c.Cmd == CommandType.Forward)
                {
                    pos += c.Num;
                    depth += c.Num * aim;
                }
                else
                {
                    aim += c.Num;
                }
            }

            Console.WriteLine("{0}p x {1}d = {2}", pos, depth, pos * depth);
        }
    }
}

File: xanathar/rs/src/day02.rs
#![allow(unused_imports)]
use std::iter::Iterator;
use std::collections::{HashMap, HashSet, VecDeque, BTreeSet, BTreeMap};
use regex::Regex;
use lazy_static::lazy_static;
use std::ops::RangeInclusive;
use crate::utils;

enum Command {
    Forward(i64),
    UpDown(i64),
}

fn parse_command(s: String) -> Option<Command> {
    let cmd: Vec<&str> = s.split(' ').collect();

    match cmd[0] {
        "forward" => Some(Command::Forward(cmd[1].parse::<i64>().unwrap())),
        "down" => Some(Command::UpDown(cmd[1].parse::<i64>().unwrap())),
        "up" => Some(Command::UpDown(-cmd[1].parse::<i64>().unwrap())),
        _ => None,
    }
}


pub fn test1() {
    let data = utils::parse_lines("../data/day2.txt", parse_command);

    let mut depth = 0i64;
    let mut pos = 0i64;

    for c in data.iter() {
        match c {
            Command::Forward(v) => pos += v,
            Command::UpDown(v) => depth += v,
        }
    }

    println!("{}", depth * pos);
}

pub fn test2() {
    let data = utils::parse_lines("../data/day2.txt", parse_command);

    let mut depth = 0i64;
    let mut aim = 0i64;
    let mut pos = 0i64;

    for c in data.iter() {
        match c {
            Command::Forward(v) => { pos += v; depth += v * aim; },
            Command::UpDown(v) => aim += v,
        }
    }

    println!("{}", depth * pos);
}


